### PR TITLE
Implement record modes for VCR.

### PR DESCRIPTION
**Pull request by abhinav**
Also add some documentation.

Note: There's an xfail of one of the VCR tests pending a fix to the patching logic.

@breerly @junchaowu @blampe

### CODE REVIEW

File: docs/api.rst [hunk 1]
```diff
@@ -40,3 +40,14 @@ Synchronous Client
 
 .. automodule:: tchannel.sync.thrift
     :members:
+
+
+Testing
```
Comment (breerly) [inline docs/api.rst]: :+1: docs :)

File: tchannel/testing/vcr/__init__.py [hunk 2]
```diff
@@ -20,10 +47,40 @@ def test_foo():
 from .cassette import Cassette
 
 
-def use_cassette(path):
+def use_cassette(path, record_mode=None):
```
Comment (breerly) [inline tchannel/testing/vcr/__init__.py]: I'm allergic to string params, can you at least provide constants?

File: tchannel/testing/vcr/__init__.py [hunk 3]
```diff
@@ -20,10 +47,40 @@ def test_foo():
...
+    """Use or create a cassette to record/replay TChannel requests.
+
+    This may be used as a context manager or a decorator.
+
+    .. code-block:: python
+
+        from tchannel.testing import vcr
+
+        @vcr.use_cassette('tests/data/foo.yaml')
```
Comment (breerly) [inline tchannel/testing/vcr/__init__.py]: I think VCR added the ability to auto-name the cassette... not sure if that's worth it or not.

File: tchannel/testing/vcr/cassette.py [hunk 4]
```diff
@@ -7,25 +7,84 @@
 from . import yaml
 from .exceptions import RequestNotFoundError
 from .exceptions import UnsupportedVersionError
+from .exceptions import VCRError
+
+
+__all__ = ['Cassette']
```
Comment (breerly) [inline tchannel/testing/vcr/cassette.py]: :+1:

File: tchannel/testing/vcr/cassette.py [hunk 5]
```diff
@@ -7,25 +7,84 @@
...
+    def __init__(self, name, replayable, can_record):
+        self.name = name
+        self.replayable = replayable
+        if not callable(can_record):
+            self.can_record = (lambda _: can_record)
+        else:
+            self.can_record = can_record
+
+
+RECORD_MODES = {m.name: m for m in [
```
Comment (breerly) [inline tchannel/testing/vcr/cassette.py]: sweet python bro

File: tchannel/testing/vcr/cassette.py [hunk 6]
```diff
@@ -7,25 +7,84 @@
...
         """Initialize a new cassette.
 
         :param path:
             File path at which this cassette will be stored.
+        :param record_mode:
+            One of 'once', 'none', 'all', 'new_episodes'. See
+            :ref:`record-modes`.
         """
+        # TODO move documentation around
+        record_mode = record_mode or 'once'
```
Comment (breerly) [inline tchannel/testing/vcr/cassette.py]: use constant instead of duplicating `once` string

File: tests/test_examples.py [hunk 7]
```diff
@@ -24,6 +24,13 @@
 
 import pytest
 
+try:
+    import __pypy__  # noqa
+except ImportError:
+    PYPY = False
+else:
+    PYPY = True
```
Comment (breerly) [inline tests/test_examples.py]: pull dis out in a new PR peez

**Comment by breerly** [comment]
Mostly nits. :shipit:

File: tests/test_examples.py (hunk 7 above)
Comment (abhinav) [inline tests/test_examples.py]: Yeah, this is on master. I'll merge from master on the vcr branch and rebase this on that.

File: tchannel/testing/vcr/cassette.py [hunk 8]
```diff
@@ -45,19 +124,23 @@ def _load(self):
...
+        if not (data and 'interactions' in data):
+            return  # file was probably empty
+
+        if int(data['version']) != VERSION:
+            raise UnsupportedVersionError(
+                'Cassette at "%s" is an unsupported version of the '
+                'format: version %s' % (self.path, str(data['version']))
+            )
+
+        self._available = [
```
Comment (junchaowu) [inline tchannel/testing/vcr/cassette.py]: try catch here incase the to_native fails.

File: tchannel/testing/vcr/patch.py [hunk 9]
```diff
@@ -46,11 +46,11 @@ def choose(group, *args, **kwargs):
             yield
 
     def __enter__(self):
+        self._exit_stack.enter_context(self.cassette)
```
Comment (junchaowu) [inline tchannel/testing/vcr/patch.py]: cool

File: tchannel/testing/vcr/cassette.py (hunk 6 above)
Comment (junchaowu) [inline tchannel/testing/vcr/cassette.py]: +1

File: tchannel/testing/vcr/server.py [hunk 10]
```diff
@@ -38,6 +39,14 @@ def handle_request(self, req, res, channel):
             res.write_body(response.body)
             return
 
+        if self.cassette.write_protected:
```
Comment (junchaowu) [inline tchannel/testing/vcr/server.py]: shall we check the write_protected before can_replay?

File: tchannel/testing/vcr/__init__.py (hunk 2 above)
Comment (abhinav) [inline tchannel/testing/vcr/__init__.py]: Fxed

File: tchannel/testing/vcr/__init__.py (hunk 3 above)
Comment (abhinav) [inline tchannel/testing/vcr/__init__.py]: We can add it later if necessary.

File: tchannel/testing/vcr/cassette.py (hunk 6 above)
Comment (abhinav) [inline tchannel/testing/vcr/cassette.py]: Fixed

File: tchannel/testing/vcr/cassette.py (hunk 8 above)
Comment (abhinav) [inline tchannel/testing/vcr/cassette.py]: If to-native fails, they probably had an invalid cassette and we want the test to blow up. It's undesirable to silently ignore failures like that during tests.

File: tchannel/testing/vcr/server.py (hunk 10 above)
Comment (abhinav) [inline tchannel/testing/vcr/server.py]: No. You can replay even if you're write protected (record mode 'none').

File: tchannel/testing/vcr/record_modes.py [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+    .. autoattribute:: NEW_EPISODES
+       :annotation: = 'new_episodes'
+
+    .. autoattribute:: NONE
+       :annotation: = 'none'
+
+    .. autoattribute:: ALL
+       :annotation: = 'all'
+    """
```
Comment (abhinav) [inline tchannel/testing/vcr/record_modes.py]: This is how this looks, BTW.

image

I wanted it to be clear that those strings can be used to refer to these record modes in addition to `RecordMode.FOO'.

**Pull request merged**